File: 1022-sum-of-root-to-leaf-binary-numbers/1022-sum-of-root-to-leaf-binary-numbers.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    int ans{0}, cur{0};
public:
    int sumRootToLeaf(TreeNode* root) {
        recurse(root);
        return ans;
    }
    void recurse(TreeNode* root){
        //shift left by 1. increasing depth
        cur = (cur << 1) | root->val;
        //test if leaf
        if(not root->left and not root->right)
            ans += cur;
        else{
            if(root->left)
                recurse(root->left);
            if(root->right)
                recurse(root->right);
        }
        //going upwards now
        cur >>= 1;
    } 
};

File: 1022-sum-of-root-to-leaf-binary-numbers/1022-sum-of-root-to-leaf-binary-numbers.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sumRootToLeaf(self, node, temp=0):
        res = 0
        if node.left == node.right == None: return (temp<<1) + node.val
        if node.left:  res += self.sumRootToLeaf(node.left, (temp<<1) + node.val)
        if node.right: res += self.sumRootToLeaf(node.right,(temp<<1) + node.val) 
        return res